README.md
# terraincache

A (yet another) basic Python script for downloading and accessing [Mapzen Terrain Tiles](https://registry.opendata.aws/terrain-tiles/) geotiffs from AWS. Designed to integrate with `rasterio` and to provide a familiar command line tool for quick reprojection of downloads.

## Installation

Installation assumes that you have `gdal` and the `gdal` Python bindings already installed to a location findable by your system.

    pip install terraincache

Set the `TERRAINCACHE` environment variable to save typing:

    export TERRAINCACHE=/users/snorris/Data/terrain-tiles

## Usage

### Python module

    from matplotlib import pyplot as plt
    import rasterio

    from terraincache import TerrainTiles

    bounds =[-125.2714, 51.3706, -125.2547, 51.3768]

    # Initialize with bounds and zoom of interest
    tt = TerrainTiles(bounds, 11)

    # load to numpy array
    array = tt.load()

    # plot the array directly
    plt.imshow(array, cmap='terrain', extent=bounds)

![alt text](dem1.png "dem1")

Dump to file and plot the array using `rasterio.plot.show()` (for correct axis labels):

    tt.save(out_file="dem.tif")
    with rasterio.open("dem.tif", "r") as src:
        fig, ax = plt.subplots()
        rasterio.plot.show(src, ax=ax, title='terrain-tiles', cmap="terrain")

![alt text](dem2.png "dem2")

### CLI

    $ terraincache --help
    Usage: terraincache [OPTIONS]

      Write terrain tiles to geotiff

    Options:
      -o, --out_file, --out-file PATH
                                      Output file name/path
      --bounds TEXT                   Bounds: "left bottom right top" or "[left,
                                      bottom, right, top]".  [required]
      -z, --zoom INTEGER              Web map zoom level  [required]
      -p, --cache-dir, --cache_dir TEXT
                                      Path to cache folder
      -tr, --resolution FLOAT         Output dataset target resolution in meters
                                      (square pixels)
      --bounds-crs, --bounds_crs TEXT
                                      CRS of provided bounds
      --dst-crs, --dst_crs TEXT       Target coordinate reference system.
      -r, --resampling_algorithm [bilinear|nearest|cubic|cubicspline|lanczos|average|mode]
                                      GDAL resampling algorithm
      -v, --verbose                   Increase verbosity.
      -q, --quiet                     Decrease verbosity.
      --help                          Show this message and exit.

For example, create a geotiff around the summit of Mt Fairweather (at BC / AK border) that aligns with the BC TRIM DEM.  Note that this is almost certainly upsampling the data available, see below link to review source resolutions.

    terraincache \
      --bounds "336637.5, 1597112.5, 346637.5, 1607112.5" \
      --bounds-crs EPSG:3005 \
      --zoom 11 \
      -tr 25 \
      -r bilinear \
      -o dem.tif


## Data sources

See the [data sources reference document](https://github.com/tilezen/joerd/blob/master/docs/data-sources.md)

## Credits / other implementations

- [terrain-tiles](https://registry.opendata.aws/terrain-tiles)
- [Mapzen provided script](https://github.com/tilezen/joerd/blob/master/docs/examples/collect.py)
- [planetutils](https://github.com/interline-io/planetutils)
- [dem](https://github.com/dgketchum/dem)
- [elevatr](https://github.com/jhollist/elevatr)

CHANGES.txt
v0.0.1 (November 2019)
----------------------
initial release

tests/test_mod.py
import os
from pathlib import Path

import mercantile
from mercantile import Tile

from terraincache import TerrainTiles


TEST_TILE = Tile(x=622, y=1364, z=12)
TEST_BOUNDS = (-125.271412, 51.370639, -125.254793, 51.376881)
TEST_ZOOM_1 = 12
TEST_ZOOM_2 = 14


def tile_path(tile):
    return "/".join([str(tile.z), str(tile.x), str(tile.y) + ".tif"])


def test_download_single_tile(tmpdir):
    tt = TerrainTiles(TEST_BOUNDS, TEST_ZOOM_1, cache_dir=str(tmpdir))
    tt.download_tile(TEST_TILE)
    assert Path(str(tmpdir)).joinpath(tile_path(TEST_TILE)).exists()


def test_download_bounds(tmpdir):
    tt = TerrainTiles(TEST_BOUNDS, TEST_ZOOM_2, cache_dir=str(tmpdir))
    for tile in mercantile.tiles(*TEST_BOUNDS, TEST_ZOOM_2):
        assert Path(str(tmpdir)).joinpath(tile_path(tile)).exists()


def test_merge(tmpdir):
    tt = TerrainTiles(TEST_BOUNDS, TEST_ZOOM_2, cache_dir=str(tmpdir))
    tt.merge(out_file=os.path.join(str(tmpdir), "merged.tif"))
    assert (Path(str(tmpdir)) / "merged.tif").exists


def test_warp(tmpdir):
    tt = TerrainTiles(
        TEST_BOUNDS,
        TEST_ZOOM_2,
        cache_dir=str(tmpdir),
        dst_crs="EPSG:3005",
        resolution=50,
    )
    tt.merge()
    tt.warp(out_file=os.path.join(str(tmpdir), "warped.tif"))
    assert (Path(str(tmpdir)) / "warped.tif").exists


def test_load(tmpdir):
    tt = TerrainTiles(
        TEST_BOUNDS,
        TEST_ZOOM_2,
        cache_dir=str(tmpdir),
        dst_crs="EPSG:3005",
        resolution=50,
    )
    array = tt.load()
    assert array.shape == (14, 23)


terraincache/terraincache.py
import os
from pathlib import Path
import logging
import tempfile
import subprocess

from osgeo import gdal
import mercantile
import rasterio
from rasterio.warp import transform_bounds

import requests


LOG = logging.getLogger(__name__)


class TerrainTiles(object):
    def __init__(
        self,
        bounds,
        zoom=11,
        bounds_crs="EPSG:4326",
        dst_crs="EPSG:4326",
        resolution=None,
        cache_dir=None,
        resampling="bilinear"
    ):

        # create a temp folder for vrts
        self.tempfolder = tempfile.mkdtemp(prefix="terrain-tiles-")

        # zoom must be 1-15
        if type(zoom) != int or zoom < 1 or zoom > 15:
            raise ValueError("Zoom must be an integer from 1-15")

        # if cache is not specified
        if not cache_dir:
            # first look for environment variable
            if "TERRAINCACHE" in os.environ.keys():
                cache_dir = os.environ["TERRAINCACHE"]
            # then default to temp folder
            else:
                cache_dir = self.tempfolder

        self.cache_dir = cache_dir
        self.zoom = zoom
        self.bounds = bounds
        if bounds_crs != "EPSG:4326":
            self.bounds_ll = transform_bounds(bounds_crs, "EPSG:4326", *bounds)
        else:
            self.bounds_ll = bounds
        self.bounds_crs = bounds_crs

        self.url = "http://s3.amazonaws.com/elevation-tiles-prod/geotiff/"
        self.tiles = [t for t in mercantile.tiles(*self.bounds_ll, self.zoom)]
        self.files = []
        self.merged = False
        self.warped = False
        self.resolution = resolution
        self.dst_crs = dst_crs
        self.resampling = resampling
        self.cache()

    def download_tile(self, tile):
        """Download a terrain tile to cache, where tile is a Mercantile Tile
        """
        tilepath = "/".join([str(tile.z), str(tile.x), str(tile.y) + ".tif"])
        LOG.debug(f"Downloading tile {tilepath} to cache")
        cachepath = Path(self.cache_dir).joinpath(tilepath)
        cachepath.parent.mkdir(parents=True, exist_ok=True)
        url = self.url + tilepath
        r = requests.get(url, stream=True)
        if r.status_code != 200:
            raise RuntimeError("No such tile: {}".format(tilepath))
        with cachepath.open("wb") as fd:
            for chunk in r.iter_content(chunk_size=128):
                fd.write(chunk)
        self.files.append(str(cachepath))

    def cache(self):
        """Find geotiffs that intersect provided bounds in cache or on web
        """
        for tile in self.tiles:
            tilepath = "/".join([str(tile.z), str(tile.x), str(tile.y) + ".tif"])
            cachepath = Path(self.cache_dir).joinpath(tilepath)
            if cachepath.exists():
                LOG.debug(f"Found tile {tilepath} in cache")
                self.files.append(str(cachepath))
            else:
                self.download_tile(tile)

    def merge(self, out_file=None):
        """
        Create virtual merge of individual terrain tiles at given zoom, bounds

        Merging with rasterio.merge may run up on system limits, just use gdal.
        https://github.com/mapbox/rasterio/issues/1636
        https://github.com/smnorris/terraincache/issues/2
        """
        self.merged = os.path.join(self.tempfolder, "merge.vrt")
        vrt_options = gdal.BuildVRTOptions(resampleAlg=self.resampling)
        gdal.BuildVRT(self.merged, self.files, options=vrt_options)
        if out_file:
            LOG.info(f"writing {out_file}")
            cmd = ["gdal_translate", "-ot", "Int16", self.merged, out_file]
            subprocess.run(cmd)
            self.merged = out_file

    def warp(self, out_file=None):
        """reproject, resample, convert to int16
        """
        if not out_file:
            out_file = os.path.join(str(self.tempfolder), "warped.vrt")
        if self.bounds_crs != self.dst_crs:
            target_extent = transform_bounds(
                self.bounds_crs, self.dst_crs, *self.bounds
            )
            te = [str(t) for t in target_extent]
        else:
            te = [str(t) for t in self.bounds]
        cmd = [
            "gdalwarp",
            "-t_srs",
            self.dst_crs,
            "-te",
            te[0],
            te[1],
            te[2],
            te[3],
            "-r",
            self.resampling,
            "-ot",
            "Int16",
            "-dstnodata",
            "-32768",
            self.merged,
            out_file,
        ]
        if self.resolution:
            cmd.extend(["-tr", str(self.resolution), str(self.resolution)])
        LOG.debug(" ".join(cmd))
        subprocess.run(cmd)
        self.warped = str(out_file)

    def load(self):
        """read merged, warped raster and return np array
        """
        if not self.merged:
            self.merge()
        if not self.warped:
            self.warp()
        with rasterio.open(self.warped) as src:
            array = src.read(1)
        return array

    def save(self, out_file):
        """save merged and warped raster to specified path
        """
        if not self.merged:
            self.merge()
        self.warp(out_file)
        self.out_file = out_file


terraincache/cli.py
import click
import logging
import sys

from terraincache import TerrainTiles
from rasterio.rio import options
from cligj import verbose_opt, quiet_opt


LOG = logging.getLogger(__name__)


@click.command("terraincache")
@click.option(
    "--out_file",
    "--out-file",
    "-o",
    default="terrain-tiles.tif",
    type=click.Path(resolve_path=True),
    help="Output file name/path"
)
@click.option(
    "--bounds",
    callback=options.bounds_handler,
    required=True,
    help='Bounds: "left bottom right top" or "[left, bottom, right, top]".',
)
@click.option("--zoom", "-z", type=int, required=True, help="Web map zoom level")
@click.option("--cache-dir", "--cache_dir", "-p", envvar="TERRAINCACHE", help="Path to cache folder")
@click.option(
    "-tr",
    "--target_resolution",
    type=float,
    default=25,
    help="Output dataset target resolution in meters (square pixels)",
)
@click.option(
    "--bounds-crs", "--bounds_crs", help="CRS of provided bounds", default="EPSG:4326"
)
@click.option(
    "--dst-crs",
    "--dst_crs",
    default="EPSG:3005",
    help="Target coordinate reference system.",
)
@click.option(
    "-r",
    "--resampling_algorithm",
    type=click.Choice(["bilinear", "nearest", "cubic", "cubicspline", "lanczos", "average", "mode"], case_sensitive=False),
    default="bilinear",
    help="GDAL resampling algorithm",
)
@verbose_opt
@quiet_opt
def cli(out_file, bounds, zoom, cache_dir, target_resolution, bounds_crs, dst_crs, resampling_algorithm, verbose, quiet):
    """Write terrain tiles to geotiff"""
    verbosity = verbose - quiet
    log_level = max(10, 30 - 10 * verbosity)
    logging.basicConfig(stream=sys.stderr, level=log_level)
    LOG.info(bounds)
    tt = TerrainTiles(
        bounds,
        zoom,
        bounds_crs=bounds_crs,
        cache_dir=cache_dir,
        resolution=target_resolution,
        dst_crs=dst_crs,
    )
    tt.save(out_file)


if __name__ == "__main__":
    cli()
